File: C++/323. Number of Connected Components in an Undirected Graph.cpp
class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        int ans = 0;
        auto g = vector<vector<int>> (n, vector<int>());
        for(auto &v : edges) {
            auto &a = v.front(), &b = v.back();
            g[a].push_back(b);
            g[b].push_back(a);
        }

        unordered_set<int> seen;
        auto dfs = [&](this auto &&dfs, int u) -> void {
            seen.insert(u);
            for(auto &e : g[u]){
                if(seen.contains(e)) continue;
                dfs(e);
            }
        };

        for(size_t i = 0; i < n; ++i) {
            if(!seen.contains(i)) {
                ans++;
                dfs(i);
            }
        }

        return ans;
    }
};

File: Swift/323. Number of Connected Components in an Undirected Graph.swift
class Solution {
    func countComponents(_ n: Int, _ edges: [[Int]]) -> Int {
        var g: [[Int]] = .init(repeating: [], count: n)
        for e in edges {
            if let a = e.first, let b = e.last {
                g[a].append(b)
                g[b].append(a)
            }
        }

        var seen = Set<Int>()
        func dfs(_ root: Int) {
            seen.insert(root)
            for x in g[root] {
                if !seen.contains(x) {
                    dfs(x)
                }
            }
        }

        var ans = 0
        for i in 0 ..< n {
            if !seen.contains(i) {
                ans += 1
                dfs(i)
            }
        }
        return ans
    }
}
